pkg/proxy/healthcheck/healthcheck.go

```diff
@@ -20,108 +20,214 @@ import (
+	if listener == nil {
+		listener = stdNetListener{}
+	}
+	if httpServerFactory == nil {
+		httpServerFactory = stdHTTPServerFactory{}
+	}
+	return &server{
+		hostname:    hostname,
+		recorder:    recorder,
+		listener:    listener,
+		httpFactory: httpServerFactory,
+		services:    map[types.NamespacedName]*hcInstance{},
+		endpoints:   map[types.NamespacedName]int{},
+	}
 }
 
-// proxyHC: Handler structure for health check, endpoint add/delete and service listener add/delete requests
-type proxyHC struct {
-	serviceEndpointsMap    cache.ThreadSafeStore
-	serviceResponderMap    map[types.NamespacedName]serviceResponder
-	mutationRequestChannel chan *proxyMutationRequest
-	listenerRequestChannel chan *proxyListenerRequest
+// Implement Listener in terms of net.Listen.
+type stdNetListener struct{}
+
+func (stdNetListener) Listen(addr string) (net.Listener, error) {
+	return net.Listen("tcp", addr)
 }
 
-// handleHealthCheckRequest - received a health check request - lookup and respond to HC.
-func (h *proxyHC) handleHealthCheckRequest(rw http.ResponseWriter, serviceName string) {
-	s, ok := h.serviceEndpointsMap.Get(serviceName)
-	if !ok {
-		glog.V(4).Infof("Service %s not found or has no local endpoints", serviceName)
-		sendHealthCheckResponse(rw, http.StatusServiceUnavailable, "No Service Endpoints Found")
-		return
-	}
-	numEndpoints := len(*s.(*serviceEndpointsList).endpoints)
-	if numEndpoints > 0 {
-		sendHealthCheckResponse(rw, http.StatusOK, fmt.Sprintf("%d Service Endpoints found", numEndpoints))
-		return
+var _ Listener = stdNetListener{}
+
+// Implement HTTPServerFactory in terms of http.Server.
+type stdHTTPServerFactory struct{}
+
+func (stdHTTPServerFactory) New(addr string, handler http.Handler) HTTPServer {
+	return &http.Server{
+		Addr:    addr,
+		Handler: handler,
 	}
-	sendHealthCheckResponse(rw, http.StatusServiceUnavailable, "0 local Endpoints are alive")
 }
 
-// handleMutationRequest - receive requests to mutate the table entry for a service
-func (h *proxyHC) handleMutationRequest(req *proxyMutationRequest) {
-	numEndpoints := len(*req.endpointUids)
-	glog.V(4).Infof("LB service health check mutation request Service: %s - %d Endpoints %v",
-		req.serviceName, numEndpoints, (*req.endpointUids).List())
-	if numEndpoints == 0 {
-		if _, ok := h.serviceEndpointsMap.Get(req.serviceName.String()); ok {
-			glog.V(4).Infof("Deleting endpoints map for service %s, all local endpoints gone", req.serviceName.String())
-			h.serviceEndpointsMap.Delete(req.serviceName.String())
+var _ HTTPServerFactory = stdHTTPServerFactory{}
+
+type server struct {
+	hostname    string
+	recorder    record.EventRecorder // can be nil
+	listener    Listener
+	httpFactory HTTPServerFactory
+
+	lock      sync.Mutex
+	services  map[types.NamespacedName]*hcInstance
+	endpoints map[types.NamespacedName]int // number of local endpoints for a service
```

Comment:
Why not combine `endpoints` with `services`? They seems to have the same entries.